lib/jsonapi/plugs/transform_parameters.ex

```diff
@@ -0,0 +1,41 @@
+defmodule JSONAPI.TransformParameters do
+  @moduledoc """
+  Takes dasherized JSON:API params and turns them into underscored params. Add
+  this to your API's pipeline to aid in dealing with incoming parameters.
+
+  ## Example
+
+  Given a request like:
+
+      GET /example?filters[dog-breed]=Corgi
+
+  You could implement your corresponding `index` method like:
+
+      def index(conn, %{"filters" => %{"dog_breed" => "Corgi"}})
+
+  Without this Plug your index action would look like:
+
+      def index(conn, %{"filters" => %{"dog-breed" => "Corgi"}})
+  """
+
+  import Plug.Conn
+
+  import JSONAPI.Utils.Underscore, only: [dasherize_keys: 1]
+
+  @json_api_content_type "application/vnd.api+json"
+
+  def init(opts), do: opts
+
+  def call(%Plug.Conn{params: params} = conn, _opts) do
+    content_type = get_req_header(conn, "content-type")
+
+    if @json_api_content_type in content_type do
+      new_params = dasherize_keys(params)
```

Comment:
My original intention was to just handle filter params, but I think we can consolidate down into a single Plug. I still haven't removed the duplication present in the query parser plug, and I _think_ we ought to (though this will be a breaking change if the plug introduced in this PR remains optional unless we can bundle them???).

Reply:
@jherdman I think opt in behavior would be a minor bump.  Is this an incompatible API change for all users?  I think this is technically backwards compat with added functionality.  Feel free to correct me!